frontend/deb/debroot.go

```diff
@@ -0,0 +1,226 @@
+
+	states := []llb.State{control, rules, changelog, debian}
+	states = append(states, installers...)
+
+	dalecDir := base.
+		File(llb.Mkdir(filepath.Join(dir, "dalec"), 0o755), opts...)
+
+	states = append(states, dalecDir.File(llb.Mkfile(filepath.Join(dir, "dalec/build.sh"), 0o700, createBuildScript(spec)), opts...))
+	states = append(states, dalecDir.File(llb.Mkfile(filepath.Join(dir, "dalec/patch.sh"), 0o700, createPatchScript(spec)), opts...))
+	states = append(states, dalecDir.File(llb.Mkfile(filepath.Join(dir, "dalec/fix_file_backed_sources.sh"), 0o700, fixupFileBackedSourcesScript(spec)), opts...))
+
+	patchDir := dalecDir.File(llb.Mkdir(filepath.Join(dir, "dalec/patches"), 0o755), opts...)
+	sorted := dalec.SortMapKeys(spec.Patches)
+	for _, name := range sorted {
+		pls, err := sourcePatchesDir(sOpt, patchDir, filepath.Join(dir, "dalec/patches"), name, spec, opts...)
+		if err != nil {
+			return llb.Scratch(), errors.Wrap(err, "error creating patch directory")
+		}
+		states = append(states, pls...)
+	}
+
+	return dalec.MergeAtPath(in, states, "/"), nil
+}
+
+func fixupFileBackedSourcesScript(spec *dalec.Spec) []byte {
+	// first, don't bother doing this for patches, which we don't include in our source tarballs.
+	patches := map[string]struct{}{}
+	for _, patchList := range spec.Patches {
+		for _, patch := range patchList {
+			patches[patch.Source] = struct{}{}
+		}
+	}
+
+	buf := bytes.NewBuffer(nil)
+	buf.WriteString("#!/usr/bin/env sh\n\n")
+	buf.WriteString("set -ex\n\n")
+
+	// now, we need to find all the sources that are file-backed and fix them up
+	for name, src := range spec.Sources {
+		if _, ok := patches[name]; ok {
+			continue
+		}
+		if dalec.SourceIsDir(src) {
+			continue
+		}
+
+		fmt.Fprintf(buf, "mv %q/%q %q.dalec.tmp\n", name, name, name)
+		fmt.Fprintln(buf, "rm -rf", name)
+		fmt.Fprintf(buf, "mv %q.dalec.tmp %q\n", name, name)
+		fmt.Fprintln(buf)
+	}
+
+	return buf.Bytes()
+}
+
+func createPatchScript(spec *dalec.Spec) []byte {
+	buf := bytes.NewBuffer(nil)
+
+	fmt.Fprintln(buf, "#!/usr/bin/env sh")
+	fmt.Fprintln(buf)
+
+	// Enable error checking to simplify the main part of the script
+	fmt.Fprintln(buf, "set -ex")
+
+	for name, patches := range spec.Patches {
+		for _, patch := range patches {
+			p := filepath.Join("${DEBIAN_DIR:=debian}/dalec/patches", name, patch.Source)
+			fmt.Fprintf(buf, "patch -d %q -p%d -s < %q\n", name, *patch.Strip, p)
+		}
+	}
+
+	return buf.Bytes()
+}
+
+func createBuildScript(spec *dalec.Spec) []byte {
+	buf := bytes.NewBuffer(nil)
+
+	fmt.Fprintln(buf, "#!/usr/bin/env sh")
+	fmt.Fprintln(buf)
+
+	fmt.Fprintln(buf, "set -ex")
+
+	for k, v := range spec.Build.Env {
+		fmt.Fprintln(buf, "export", k+"="+v)
+	}
+
+	for _, step := range spec.Build.Steps {
+		fmt.Fprintln(buf)
+		fmt.Fprintln(buf, "(")
+
+		for k, v := range step.Env {
+			fmt.Fprintf(buf, "	export %s=%s\n", k, v)
```

Comment:
Handle spaces in the value.
```suggestion
			fmt.Fprintf(buf, "	export %s=\"%s\"\n", k, v)
```
or it could be handle with %q not super sure on how it behaves
```suggestion
			fmt.Fprintf(buf, "	export %s=%q\n", k, v)
```

Reply:
DOne.